label/service.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func NewService(st *Store, urmCreator UserResourceMappingCreator) influxdb.LabelService {
+	return &Service{
+		store:      st,
+		urmCreator: urmCreator, // todo (al) this can be removed once URMs are removed from the Label service
+	}
+}
+
+// CreateLabel creates a new label.
+func (s *Service) CreateLabel(ctx context.Context, l *influxdb.Label) error {
+	if err := l.Validate(); err != nil {
+		return &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Err:  err,
+		}
+	}
+
+	l.Name = strings.TrimSpace(l.Name)
+
+	err := s.store.Update(ctx, func(tx kv.Tx) error {
+		if err := uniqueLabelName(ctx, tx, l); err != nil {
+			return err
+		}
+
+		if err := s.store.CreateLabel(ctx, tx, l); err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+	err = s.store.Update(ctx, func(tx kv.Tx) error {
+		if err := s.urmCreator.CreateUserResourceMappingForOrg(ctx, tx, l.OrgID, l.ID, influxdb.LabelsResourceType); err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+// FindLabelByID finds a label by its ID
+func (s *Service) FindLabelByID(ctx context.Context, id influxdb.ID) (*influxdb.Label, error) {
+	var l *influxdb.Label
+
+	err := s.store.View(ctx, func(tx kv.Tx) error {
+		label, e := s.store.GetLabel(ctx, tx, id)
+		if e != nil {
+			return e
+		}
+		l = label
+		return nil
+	})
+
+	if err != nil {
+		return nil, &influxdb.Error{
+			Err: err,
+		}
+	}
+
+	return l, nil
+}
+
+// FindLabels returns a list of labels that match a filter.
+func (s *Service) FindLabels(ctx context.Context, filter influxdb.LabelFilter, opt ...influxdb.FindOptions) ([]*influxdb.Label, error) {
+	ls := []*influxdb.Label{}
+	err := s.store.View(ctx, func(tx kv.Tx) error {
+		labels, err := s.store.ListLabels(ctx, tx, filter)
+		if err != nil {
+			return err
+		}
+		ls = labels
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return ls, nil
+}
+
+func (s *Service) FindResourceLabels(ctx context.Context, filter influxdb.LabelMappingFilter) ([]*influxdb.Label, error) {
+	ls := []*influxdb.Label{}
+	if err := s.store.View(ctx, func(tx kv.Tx) error {
+		return s.store.FindResourceLabels(ctx, tx, filter, &ls)
+	}); err != nil {
+		return nil, err
+	}
+
+	return ls, nil
+}
+
+// UpdateLabel updates a label.
+func (s *Service) UpdateLabel(ctx context.Context, id influxdb.ID, upd influxdb.LabelUpdate) (*influxdb.Label, error) {
+	var label *influxdb.Label
+	err := s.store.View(ctx, func(tx kv.Tx) error {
+		l, err := s.store.GetLabel(ctx, tx, id)
+		if err != nil {
+			return err
+		}
+		label = l
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(upd.Properties) > 0 && label.Properties == nil {
+		label.Properties = make(map[string]string)
+	}
+
+	for k, v := range upd.Properties {
+		if v == "" {
+			delete(label.Properties, k)
+		} else {
+			label.Properties[k] = v
+		}
+	}
+
+	if upd.Name != "" {
+		err := s.store.Update(ctx, func(tx kv.Tx) error {
```

Comment:
I feel like this logic should live in Store and the `*label.Store.UpdateLabel` should take the LabelUpdate.
Meaning all Service does is open a transaction and pass the right things down to storage.
My concern here is that after updating the name, we could fail validation of the label or fail to actually put the label, which is all `store.UpdateLabel` is doing at the moment.
This would also reduce this function down to 1 transaction instead of opening 3.
So you could get a non-successfull call to this service to update the label, but the name did change successfully.